src/Device/Driver/LX_EOS/LXEosDownload.cpp

```diff
@@ -0,0 +1,245 @@
+#include "Device/Error.hpp"
+#include "Device/Port/Port.hpp"
+#include "Device/RecordedFlight.hpp"
+#include "LXEosDevice.hpp"
+#include "Operation/Operation.hpp"
+#include "io/BufferedOutputStream.hxx"
+#include "io/FileOutputStream.hxx"
+#include "system/Path.hpp"
+#include "util/AllocatedArray.hxx"
+#include "util/ByteOrder.hxx"
+#include "util/CRC.hpp"
+#include "util/ScopeExit.hxx"
+
+bool
+LXEosDevice::ReadFlightList(RecordedFlightList& flight_list,
+                            OperationEnvironment& env)
+{
+  port.StopRxThread();
+
+  uint8_t flight_count = 0;
+
+  env.SetProgressRange(1);
+  env.SetProgressPosition(0);
+
+  bool success = GetNumberOfFlights(flight_count, env);
+
+  // Maximum number of flights to be obtained is limited by length of the list
+  if (flight_count > flight_list.size())
+    flight_count = flight_list.size();
+
+  env.SetProgressRange(flight_count + 1);
+  env.SetProgressPosition(1);
+
+  if (success) {
+    for (uint8_t i = 0; i < flight_count; i++) {
+      RecordedFlightInfo flight;
+
+      // Try up to 5 times before giving up
+      for (uint8_t attempt = 0; attempt < 5; attempt++) {
+        success = GetFlightInfo(i + 1, flight, env);
+        if (success)
+          break;
+      }
+
+      if (!success)
+        break;
+
+      flight_list.append(flight);
+      env.SetProgressPosition(i + 2);
+    }
+  }
+
+  env.SetProgressPosition(flight_count + 1);
+
+  port.StartRxThread();
+  return success;
+}
+
+bool
+LXEosDevice::DownloadFlight(const RecordedFlightInfo& flight,
+                            Path path,
+                            OperationEnvironment& env)
+{
+  FileOutputStream fos(path);
+  BufferedOutputStream bos(fos);
+
+  bool success = true;
+
+  port.StopRxThread();
+
+  uint16_t flight_id = flight.internal.lx_eos.flight_id;
+  uint32_t bytes_remaining = flight.internal.lx_eos.file_size;
+
+  env.SetProgressRange(100);
+  env.SetProgressPosition(0);
+
+  // Download blocks until remaining bytes reach zero
+  for (uint16_t block_id = 0; bytes_remaining > 0; block_id++) {
+    AllocatedArray<std::byte> block;
+
+    // Try up to 5 times before giving up
+    for (uint8_t attempt = 0; attempt < 5; attempt++) {
+      try {
+        block = GetFlightLogBlock(flight_id, block_id, env);
+      } catch (...) {
+        success = false;
```

Comment:
This discards details about the error condition. Reporting error details to the user (or the log file) is very useful, don't throw that away!